File: 1285-balance-a-binary-search-tree/balance-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def balanceBST(self, root):
        def inorder(node):
            if not node:
                return []
            return inorder(node.left) + [node.val] + inorder(node.right)
        
        def array_to_tree(l, r):
            if l > r:
                return None
            mid = (l + r) // 2
            node = TreeNode(inorder[mid])
            node.left = array_to_tree(l, mid - 1)
            node.right = array_to_tree(mid + 1, r)
            return node
        
        inorder = inorder(root)
        return array_to_tree(0, len(inorder) - 1)

File: 1285-balance-a-binary-search-tree/balance-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode balanceBST(TreeNode root) {
        List<Integer> inorder = inorder(root);
        return arrayToTree(inorder, 0, inorder.size() - 1);
    }

    private List<Integer> inorder(TreeNode node) {
        List<Integer> result = new ArrayList<>();
        if (node == null) {
            return result;
        }
        result.addAll(inorder(node.left));
        result.add(node.val);
        result.addAll(inorder(node.right));
        return result;
    }

    private TreeNode arrayToTree(List<Integer> inorder, int left, int right) {
        if (left > right) {
            return null;
        }
        int mid = (left + right) / 2;
        TreeNode node = new TreeNode(inorder.get(mid));
        node.left = arrayToTree(inorder, left, mid - 1);
        node.right = arrayToTree(inorder, mid + 1, right);
        return node;
    }
}

